core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -2622,21 +2644,34 @@ private static class VariableCollector extends RexVisitorImpl<Void> {
 
   /** Represents a simple Comparison.
    *
-   * <p>Left hand side is a {@link RexNode}, right hand side is a literal.
+   * <p>Left hand side is a {@link RexNode}, right hand side is a literal/array-call.
    */
   private static class Comparison implements Predicate {
     final RexNode ref;
     final SqlKind kind;
+    //RexCall or RexLiteral
```

Comment:
It might be better to change the `RexLiteral literal` field to `RexNode argument`, and not create a separate `RexCall rexCall` field. If there's one field we can check that it is not null.